File: crates/pulsing-forge/src/discovery/marketplace.rs
//! Codex marketplace.json discovery (aligned with python/pulsing/forge/discovery/marketplace.py).

use std::path::{Path, PathBuf};

use serde_json::Value;

use super::codex_paths::{discover_marketplace_roots, find_marketplace_manifest};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum InstallPolicy {
    NotAvailable,
    Available,
    InstalledByDefault,
}

#[derive(Clone, Debug)]
pub struct MarketplacePluginSource {
    pub kind: String,
    pub local_path: Option<PathBuf>,
}

#[derive(Clone, Debug)]
pub struct MarketplacePluginEntry {
    pub name: String,
    pub source: MarketplacePluginSource,
    pub installation: InstallPolicy,
    pub marketplace_name: String,
    pub marketplace_root: PathBuf,
}

impl MarketplacePluginEntry {
    pub fn plugin_id(&self) -> String {
        format!("{}@{}", self.name, self.marketplace_name)
    }
}

#[derive(Clone, Debug)]
pub struct Marketplace {
    pub name: String,
    pub root: PathBuf,
    pub manifest_path: PathBuf,
    pub plugins: Vec<MarketplacePluginEntry>,
}

pub fn list_marketplaces(extra_roots: &[PathBuf]) -> Vec<Marketplace> {
    let mut out = Vec::new();
    for root in discover_marketplace_roots(extra_roots) {
        let Some(manifest_path) = find_marketplace_manifest(&root) else {
            continue;
        };
        if let Ok(marketplace) = load_marketplace(&manifest_path) {
            out.push(marketplace);
        }
    }
    out
}

pub fn load_marketplace(manifest_path: &Path) -> Result<Marketplace, String> {
    let raw = std::fs::read_to_string(manifest_path).map_err(|e| e.to_string())?;
    let value: Value = serde_json::from_str(&raw).map_err(|e| e.to_string())?;
    let obj = value.as_object().ok_or_else(|| {
        format!(
            "{}: marketplace must be a JSON object",
            manifest_path.display()
        )
    })?;
    let name = obj.get("name").and_then(|v| v.as_str()).unwrap_or_else(|| {
        manifest_path
            .parent()
            .and_then(|p| p.file_name())
            .and_then(|s| s.to_str())
            .unwrap_or("marketplace")
    });
    let marketplace_root =
        if manifest_path.file_name().and_then(|s| s.to_str()) == Some("marketplace.json") {
            manifest_path
                .parent()
                .unwrap_or(manifest_path)
                .to_path_buf()
        } else {
            marketplace_root_from_manifest(manifest_path)
        };
    let mut plugins = Vec::new();
    if let Some(items) = obj.get("plugins").and_then(|v| v.as_array()) {
        for item in items {
            let Some(entry_obj) = item.as_object() else {
                continue;
            };
            let plugin_name = entry_obj
                .get("name")
                .and_then(|v| v.as_str())
                .unwrap_or("")
                .trim()
                .to_string();
            if plugin_name.is_empty() {
                continue;
            }
            let policy_raw = entry_obj
                .get("policy")
                .and_then(|v| v.as_object())
                .and_then(|p| p.get("installation"))
                .and_then(|v| v.as_str())
                .unwrap_or("AVAILABLE");
            let installation = match policy_raw {
                "NOT_AVAILABLE" => InstallPolicy::NotAvailable,
                "INSTALLED_BY_DEFAULT" => InstallPolicy::InstalledByDefault,
                _ => InstallPolicy::Available,
            };
            let source = parse_source(
                entry_obj.get("source").and_then(|v| v.as_object()),
                &marketplace_root,
            );
            plugins.push(MarketplacePluginEntry {
                name: plugin_name,
                source,
                installation,
                marketplace_name: name.to_string(),
                marketplace_root: marketplace_root.clone(),
            });
        }
    }
    Ok(Marketplace {
        name: name.to_string(),
        root: marketplace_root.clone(),
        manifest_path: manifest_path.to_path_buf(),
        plugins,
    })
}

fn marketplace_root_from_manifest(manifest_path: &Path) -> PathBuf {
    for rel in super::codex_paths::MARKETPLACE_MANIFEST_RELATIVE_PATHS {
        let parts: Vec<_> = Path::new(rel).components().collect();
        let mut current = manifest_path.to_path_buf();
        let mut matched = true;
        for part in parts.iter().rev() {
            if current.file_name() != part.as_os_str().into() {
                matched = false;
                break;
            }
            current = current.parent().unwrap_or(&current).to_path_buf();
        }
        if matched {
            return current;
        }
    }
    manifest_path
        .parent()
        .unwrap_or(manifest_path)
        .to_path_buf()
}

fn parse_source(
    raw: Option<&serde_json::Map<String, Value>>,
    marketplace_root: &Path,
) -> MarketplacePluginSource {
    let Some(raw) = raw else {
        return MarketplacePluginSource {
            kind: "local".into(),
            local_path: Some(marketplace_root.to_path_buf()),
        };
    };
    let kind = raw
        .get("source")
        .and_then(|v| v.as_str())
        .unwrap_or("local")
        .to_ascii_lowercase();
    if kind == "git" {
        return MarketplacePluginSource {
            kind: "git".into(),
            local_path: None,
        };
    }
    let rel = raw.get("path").and_then(|v| v.as_str()).unwrap_or(".");
    let local = marketplace_root.join(rel);
    MarketplacePluginSource {
        kind: "local".into(),
        local_path: Some(local),
    }
}

File: python/pulsing/forge/discovery/marketplace.py
# SPDX-License-Identifier: Apache-2.0
"""Codex marketplace.json discovery (aligned with codex-rs/core-plugins/marketplace.rs)."""

from __future__ import annotations

import json
import shutil
import subprocess
import tempfile
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

from pulsing.forge.discovery.codex_manifest import load_codex_manifest
from pulsing.forge.discovery.codex_paths import (
    discover_marketplace_roots,
    find_marketplace_manifest,
)
from pulsing.forge.discovery.plugin_id import PluginId


class InstallPolicy(str, Enum):
    NOT_AVAILABLE = "NOT_AVAILABLE"
    AVAILABLE = "AVAILABLE"
    INSTALLED_BY_DEFAULT = "INSTALLED_BY_DEFAULT"


@dataclass
class MarketplacePluginSource:
    kind: str
    local_path: Path | None = None
    git_url: str | None = None
    git_path: str | None = None
    git_ref: str | None = None


@dataclass
class MarketplacePluginEntry:
    name: str
    source: MarketplacePluginSource
    installation: InstallPolicy = InstallPolicy.AVAILABLE
    marketplace_name: str = ""
    marketplace_root: Path | None = None

    @property
    def plugin_id(self) -> PluginId:
        return PluginId(plugin_name=self.name, marketplace_name=self.marketplace_name)


@dataclass
class Marketplace:
    name: str
    root: Path
    manifest_path: Path
    plugins: list[MarketplacePluginEntry] = field(default_factory=list)


def list_marketplaces(extra_roots: list[Path] | None = None) -> list[Marketplace]:
    out: list[Marketplace] = []
    for root in discover_marketplace_roots(extra_roots):
        manifest_path = find_marketplace_manifest(root)
        if manifest_path is None:
            continue
        try:
            out.append(load_marketplace(manifest_path))
        except (OSError, json.JSONDecodeError, ValueError):
            continue
    return out


def load_marketplace(manifest_path: Path) -> Marketplace:
    raw = json.loads(manifest_path.read_text(encoding="utf-8"))
    if not isinstance(raw, dict):
        raise ValueError(f"{manifest_path}: marketplace must be a JSON object")
    name = str(raw.get("name") or manifest_path.parent.name)
    marketplace_root = (
        manifest_path.parent
        if manifest_path.name == "marketplace.json"
        else _marketplace_root_from_manifest(manifest_path)
    )
    plugins: list[MarketplacePluginEntry] = []
    for item in raw.get("plugins") or []:
        if not isinstance(item, dict):
            continue
        plugin_name = str(item.get("name", "")).strip()
        if not plugin_name:
            continue
        policy_raw = (item.get("policy") or {}).get("installation", "AVAILABLE")
        try:
            installation = InstallPolicy(str(policy_raw))
        except ValueError:
            installation = InstallPolicy.AVAILABLE
        source = _parse_source(item.get("source") or {}, marketplace_root)
        plugins.append(
            MarketplacePluginEntry(
                name=plugin_name,
                source=source,
                installation=installation,
                marketplace_name=name,
                marketplace_root=marketplace_root,
            )
        )
    return Marketplace(
        name=name,
        root=marketplace_root,
        manifest_path=manifest_path,
        plugins=plugins,
    )


def find_installable_plugin(plugin_id: str) -> MarketplacePluginEntry:
    pid = PluginId.parse(plugin_id)
    for marketplace in list_marketplaces():
        if marketplace.name != pid.marketplace_name:
            continue
        for plugin in marketplace.plugins:
            if plugin.name != pid.plugin_name:
                continue
            if plugin.installation == InstallPolicy.NOT_AVAILABLE:
                raise ValueError(
                    f"plugin {plugin_id!r} is not available for install in marketplace {marketplace.name!r}"
                )
            return plugin
    raise ValueError(f"plugin {plugin_id!r} was not found in any marketplace")


def materialize_plugin_source(entry: MarketplacePluginEntry) -> Path:
    source = entry.source
    if source.kind == "local":
        if source.local_path is None or not source.local_path.is_dir():
            raise ValueError(f"local plugin source missing for {entry.plugin_id.id}")
        return source.local_path.resolve()
    if source.kind == "git":
        if not source.git_url:
            raise ValueError(f"git plugin source missing url for {entry.plugin_id.id}")
        tmp = Path(tempfile.mkdtemp(prefix="forge-plugin-src-"))
        cmd = ["git", "clone", "--depth", "1"]
        if source.git_ref:
            cmd.extend(["--branch", source.git_ref])
        cmd.extend([source.git_url, str(tmp)])
        subprocess.run(cmd, check=True, capture_output=True, text=True)
        if source.git_path:
            sub = (tmp / source.git_path).resolve()
            if not sub.is_dir():
                raise ValueError(f"git plugin subpath missing: {source.git_path}")
            return sub
        return tmp
    raise ValueError(f"unsupported plugin source kind {source.kind!r}")


def _marketplace_root_from_manifest(manifest_path: Path) -> Path:
    for rel in (".agents/plugins/marketplace.json", ".claude-plugin/marketplace.json"):
        parts = Path(rel).parts
        current = manifest_path
        for part in reversed(parts):
            if current.name != part:
                break
            current = current.parent
        else:
            return current
    return manifest_path.parent


def _parse_source(
    raw: dict[str, Any], marketplace_root: Path
) -> MarketplacePluginSource:
    kind = str(raw.get("source", "local")).lower()
    if kind == "git":
        return MarketplacePluginSource(
            kind="git",
            git_url=str(raw.get("url", "")),
            git_path=str(raw["path"]) if raw.get("path") else None,
            git_ref=str(raw.get("ref") or raw.get("sha") or "") or None,
        )
    rel = str(raw.get("path", "."))
    local = (marketplace_root / rel).resolve()
    return MarketplacePluginSource(kind="local", local_path=local)


def copy_tree_atomic(src: Path, dst: Path) -> None:
    if dst.exists():
        shutil.rmtree(dst)
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copytree(src, dst)
